freedomofpress/securedrop-client
# Implement correct export/print stages and dialog behavior

# Description

* closes #642
* Towards https://github.com/freedomofpress/securedrop-client/issues/549

A lot of messaging has been done as a group during meetings and in Slack. We still need to finalize the messaging so this can still be considered a rough draft but closer to what we want.

* Resolves https://github.com/freedomofpress/securedrop-client/issues/557

In order to do this in Qubes I had to set the window flag to `Qt.Widget` (some window flags in Qubes do not work unless you tell Qt to bypass the window manager, which is a dangerous thing to do). Since the dialog is a widget, the user can continue to interact with the client while the dialog is open. In order to avoid a race condition, I made it so only one print or export dialog can be opened at a time. Since the dialog cannot be moved, it is very unlikely that the user will try to interact with the client while it is open, but we need to make sure nothing bad happens if they do during test (see Test Plan).

This change also required adding close and cancel buttons to close the dialog since it can no longer be closed by clicking outside of it. Since the cancel button was added, I added a continue button next to it so we could see what the buttons look like together. 

* Resolves https://github.com/freedomofpress/securedrop-client/issues/618

This should have been a separate PR and can be factored out, but will take a bit of effort, so I want to offer a code walk-through first in case that helps. The speedbump was added for both print and export, which was one more reason to create a FramelessDialog class that they could both derive from. The FramelessDialog class sets the window flags and creates the layout and styles that both print and export share. The layout looks like this: 
![Screenshot from 2020-01-21 01-04-35](https://user-images.githubusercontent.com/4522213/72790266-2bb6ae80-3bea-11ea-8adf-aa6f27891274.png)

* Resolves updating the UI, see https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5ddc3b8b9d9aab94be0542d5 for reference, which has gone through many iterations

* Added filename to PrintDialog so that "Preparing to print" -> "Preparing to print: filename.pdf"
* Added "Export successful" message since we removed it from the Export VM. I created a [followup issue](https://github.com/freedomofpress/securedrop-client/issues/713) to make the color green and icon image to a checkmark.
* Added show/hide password to passphrase form (using the same eye icon as we use in the login dialog)

# Test Plan

1. Click outside of dialog for both print and export to verify it no longer closes the dialog
2. Make sure you can only open one dialog at a time, either a single print dialog or a single export dialog
3. Resize the window and see that the dialog continues to be centered with the application
4. Click on "Print" and see that filename now shows up in preparation window
5. Click on "Cancel" to verify that it closes the dialog for both print and export
6. Test show/hide password
7. See "Export successful" message when the export succeeds
8. See dialog close when the Export VM begins opening the 3rd party printer application
9. Make sure you can still export and print documents and test all the error cases: (1) bad passphrase, (2) no usb device connected, (3) multiple devices connected, (4) unencrypted usb device connected
